Cpp_test/lib/ipc-shm-send.cpp

```diff
@@ -0,0 +1,123 @@
+#include "ipc-shm-send.h"
+#include <iostream>
+#include <cassert>
+
+IPCShmSend::~IPCShmSend()
+{
+
+	shm_unlink(this->info.method_name);
+	
+}
+
+void IPCShmSend::init()
+{
+	this->file_handler.open_file();
+	this->open_shm();
+	this->set_shm_size();
+	this->map_shm();
+	if (this->shmd > 0)
+	{
+		close(this->shmd);
+	}
+	this->init_mutex();
+	this->init_cond();
+}
+
+void IPCShmSend::transfer()
+{
+	long read_bytes{0};
+	unsigned long long file_size{this->file_handler.get_file_size()};
+	if (file_size == 0)
+	{
+		throw std::runtime_error("ERROR: File size = 0.");
+	}
+
+	unsigned long long total_sent_bytes{0};
+	std::cout << "Waiting for receiver..." << std::endl;
+	while (!this->is_end)
+	{
+		lock_mutex();
+		while (this->shm_ptr->data_version_received != this->shm_ptr->data_version)
+		{
+
+			if (pthread_cond_wait(&(this->shm_ptr->cond_re), &(this->shm_ptr->mutex)) != 0)
+			{
+				throw std::runtime_error(std::string("ERROR: pthread_cond_wait() send: ") + strerror(errno));
+			}
+		}
+
+		if (total_sent_bytes < file_size)
+		{
+			this->file_handler.read_file_shm(this->shm_ptr->data_ap, this->shm_ptr->data_size);
+			read_bytes = this->file_handler.get_read_bytes();
+
+			if (read_bytes > 0)
+			{
+				this->shm_ptr->data_version++;
+				this->shm_ptr->data_size = read_bytes;
+				total_sent_bytes += read_bytes;
+			}
+		}
+		else
+		{
+			this->shm_ptr->data_size = 0;
+			this->shm_ptr->data_version++;
+			this->is_end = true;
+		}
+		send_cond_broadcast();
+		unlock_mutex();
+	}
+
+	if (total_sent_bytes == file_size && this->is_end)
+	{
+		std::cout << "Sent: " << total_sent_bytes << " byte(s)." << std::endl;
+	}
+	else
+	{
+		throw std::runtime_error("ERROR: The size of Total send file is not equal to File size.");
+	}
+	
+	if (pthread_cond_destroy(&(this->shm_ptr->cond)) != 0)
+	{
+		perror("Error at pthread_cond_destroy()\n");
+	}
+	
+	if (pthread_mutex_destroy(&(this->shm_ptr->mutex)) != 0)
+	{
+		perror("Error at pthread_mutex_destroy()\n");
+	}
```

Comment:
If you do this, the cond and the mutex will be destroy only if the transfer)_ method is called and this method doesn't throw.